### ISSUE TITLE
Add Idea Hub feature tour on WordPress posts page

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3625]
## Feature Description

In addition to #3524, another feature tour should showcase the possible Idea Hub interactions on the WordPress posts page.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

A new tooltip tour `ideaHubPostsTable` should be displayed on the WordPress admin posts list table _if_ the `ideaHubModule` feature flag is enabled. Additional requirements to show the tooltip:
* The Idea Hub module is active.
* The user has successfully completed Idea Hub setup.
* The user has created at least one draft from Idea Hub

**Tooltip**
* Position: point to the first saved draft entry on the Posts page 
* Heading: _Get started writing about your saved ideas_ 
* Body: _Drafts you created from Idea Hub within Site Kit appear here for easy access._ 

Note that this will require to load the feature tooltip infrastructure on a non-Site Kit screen for the first time. The JS files to do that should for now only be loaded if necessary (i.e. if the above conditions are fulfilled and if the above feature tour hasn't been dismissed yet).

## Implementation Brief

Our high level approach is as follows:

1.  Add a tooltip tour and a new tooltip tour type (context) for posts list.
1.  Add dependencies to our existing JS file on the `edit.php` screen to run the tooltip tour `checkRequirements().` and to render the `<FeatureTour />`
1.  Create a container for rendering a js app.
1.  Add classes to each row of a draft idea hub post so that we can attach a tour to it.
1.  Render our tooltip tours on the edit.php screen

**Before you begin**, keep in mind that you may need to further update either the javascript file or dependencies list to add any missing javascript dependencies we need for loading the functionality as you progress with the execution of this ticket.

Here's how we achieve it:

* Rename `js/googlesitekit-idea-hub-post-list-notice.js` to `js/googlesitekit-idea-hub-post-list.js`:
  * in `includes/Modules/Idea_Hub.php`
  * in `webpack.config.js`

* In `includes/Modules/Idea_Hub.php` 
  * For the script `'googlesitekit-idea-hub-post-list-notice'` (which we just renamed to `googlesitekit-idea-hub-post-list`) add to the `dependencies`: 'googlesitekit-modules-idea-hub'. This will give our selectors in the `checkRequirements` to follow.
  * In the `register()` function, in `is_connected`, add a filter for `post_class` which will be used as the anchor point for the tooltip tour - https://developer.wordpress.org/reference/hooks/post_class/
     * Confirm that the post type = post and we are on the edit-post of `get_current_screen()->id` (see the implementation of `admin_notice_idea_hub_ideas`)
     * Confirm that the post type is a draft and an idea post (see the implementation in `display_post_states` hook in the same file).
     * If both checks above are true, add the class `googlesitekit-idea-hub__draft`
  * In the `register()` function, in `is_connected`, add a hook for `admin_footer` which will be used to render the <FeatureToursDesktop /> component:
 ```php
add_action('admin_footer', function() { echo '<div id="js-googlesitekit-post-list" class="googlesitekit-page"></div>';  } );
``` 

* Edit `googlesitekit/constants.js`
  * Add a new constant `export const VIEW_CONTEXT_POSTS_LIST = 'postsList';` - this will be the context for tours on the edit posts page.

* In our renamed `js/googlesitekit-idea-hub-post-list.js` render the react app:
  * `import { render } from '@wordpress/element';`
  * `import { FeatureToursDesktop } from './components/FeatureToursDesktop';`
  * In existing `domReady()`, render the react component as follows:

```js
const renderTarget = document.getElementById( 'js-googlesitekit-post-list' );
  	render(
		<FeatureToursDesktop viewContext={ 'VIEW_CONTEXT_POSTS_LIST' } />,
		renderTarget,
	);
``` 

* Create a new file `assets/js/feature-tours/idea-hub-posts-table.js` ( For reference, you can take a look at `assets/js/feature-tours/idea-hub-module.js` from #3524 ) that exports an object with the following properties:
  * `slug` should be `ideaHubPostsTable`;
  * `contexts` should be an array with `VIEW_CONTEXT_POSTS_LIST` value;
  * `version` should be `1.36.0`; - with an @TODO - change this version if the launch version for the feature changes.
  * `checkRequirements` should check the following:
    * At least one element with class `googlesitekit-idea-hub__draft` exists.
    * Idea Hub Module is active `Data.select( CORE_MODULES ).isModuleActive( 'idea-hub' )`  (console: `window.googlesitekit.data.stores["core/modules"].selectors.isModuleActive('idea-hub')`)
    * The user has successfully completed Idea Hub setup. -  `Data.select( CORE_MODULES ).isModuleConnected( 'idea-hub' )`  (`window.googlesitekit.data.stores["core/modules"].selectors.isModuleConnected('idea-hub')`)
    * There are "draft" posts from idea hub ideas. `Data.select( MODULES_IDEA_HUB ).getDraftPostIdeas()` (`window.googlesitekit.data.stores['modules/idea-hub'].selectors.getDraftPostIdeas()`)
  * `steps` should be an array with ONE item: 

  1)
      * `target` selector: `.googlesitekit-idea-hub__draft` we added this class via the hook earlier.
      * `title` should be as per the AC `Heading`
      * `content` should be as per the AC `Body`
      * `placement`  top (this may need to be adjusted once you have the feature tour loading to figure out the optimal placement)

* Add your tour to `assets/js/feature-tours/index.js`, loading it conditionally based on whether or not the feature flag `ideaHubModule` is enabled.

Add new storybook story for the tooltip tour, reference the following as an example: https://google.github.io/site-kit-wp/storybook/main/?path=/story/global--tourtooltips

### Test Coverage

* Update backstop VRT to cover the new tooltip tour.

### Visual Regression Changes

* Added new tests as above.

## QA Brief

* Make sure you have no idea posts yet
* Go to the posts page and verify that no feature tour is displayed
* Create a new draft post using any idea
* Go to the posts page again and confirm that you see the new feature tour for your newly created post.

## Changelog entry

* Add feature tour for Idea Hub drafts to the post list table in WP admin.

**Comment by felixarntz** [comment]
@eclarke1 @fhollis Just flagging that this issue is _not_ a launch blocker for Idea Hub, and it doesn't need to be ready for the eventual bug bash. It's still something we should work on soon - if we can finish it before launching Idea Hub, that would be ideal.

**Comment by danielgent** [comment]
I understand the JS side of this (using joyride.js is fairly straightforward) but I'm not sure how we actually run React code on the admin Post page.

I'm assuming that WordPress needs to enqueue our code (or a subset of it) for this page.  Probably best if someone with more WordPress knowledge explains this to me first (or I'm happy to spend a few hours learning about this)

**Comment by ivankruchkoff** [comment]
In AC
> Note that this will require to load the feature tooltip infrastructure on a non-Site Kit screen for the first time. The JS files to do that should for now only be loaded if necessary (i.e. if the above conditions are fulfilled and if the above feature tour hasn't been dismissed yet).

I don't think we need to load the JS conditionally at this point since one of the things that came up in the PR for #3359 was how much of the existing JS we needed to load in dependencies, see this comment from the PR: https://github.com/google/site-kit-wp/pull/3668#discussion_r672117922

**Comment by eugene-manuilov** [comment]
Thanks, @ivankruchkoff. IB mostly looks good. Have a few comments. Please, check it out.

> * Rename js/googlesitekit-idea-hub-post-list-notice.js to js/googlesitekit-idea-hub-post-list.js:
>   * in includes/Core/Assets/Manifest.php

We don't need to rename it in the Manifest class because it is auto-generated and will use the correct filename next time you run the `npm run build:dev` command.

> `add_action('admin_footer', function() { echo '<div id="js-googlesitekit-post-list" class="googlesitekit-page" />'; });`

`<div>`s are not valid self-closing tags. It should be `echo '<div ...></div>'`.

> * checkRequirements should check the following:
>   * Idea Hub Module is active ["core/modules"].selectors.isModuleActive('idea-hub')
>   * The user has successfully completed Idea Hub setup. - ["core/modules"].selectors.isModuleConnected('idea-hub')
>   * There are "draft" posts from idea hub ideas. ['modules/idea-hub'].selectors.getDraftPostIdeas()

We need to use `Data.select( CORE_MODULES ).isModuleActive( 'idea-hub' )` instead of `["..."].selectors...`.

**Comment by felixarntz** [comment]
@ivankruchkoff 
> I don't think we need to load the JS conditionally at this point since one of the things that came up in the PR for #3359 was how much of the existing JS we needed to load in dependencies

I'm not sure I understand - maybe it's unclear what I meant. When I say conditionally, I mean the JS file should only be enqueued (from the server-side) if it's actually needed. In other words, if that Idea Hub feature tour is already dismissed or if Idea Hub isn't even active (both of which we can check on the server-side), we shouldn't enqueue any JS.

**Comment by ivankruchkoff** [comment]
> @ivankruchkoff
> ... if that Idea Hub feature tour is already dismissed or if Idea Hub isn't even active (both of which we can check on the server-side), we shouldn't enqueue any JS.

So do we want a separate JS entry point from the work on #3359 ( https://github.com/google/site-kit-wp/blob/7aeb45dca8a631936aaf401c85802543c3891adf/assets/js/googlesitekit-idea-hub-post-list-notice.js )

That is, do we want one file to handle feature tours, and one to handle admin notices? 
It seems more straightforward to have one file for the `/wp-admin/edit.php` page, enqueued when idea hub is active, containing all idea hub functionality for that page: notice dismissal, feature tour, etc

**Comment by eugene-manuilov** [comment]
> `checkRequirements` should check the following

@ivankruchkoff I think the last thing we need to add is a check to make sure that at least one element with the `.googlesitekit-idea-hub__draft` class exists.

> That is, do we want one file to handle feature tours, and one to handle admin notices?
> It seems more straightforward to have one file for the /wp-admin/edit.php page, enqueued when idea hub is active, containing all idea hub functionality for that page: notice dismissal, feature tour, etc

IMO one file is better knowing that we load it anyway.

**Comment by eugene-manuilov** [comment]
IB :heavy_check_mark:

**Comment by eugene-manuilov** [comment]
@felixarntz which ga event category do we need to use for this feature tour?

**Comment by marrrmarrr** [comment]
@eugene-manuilov these appear all within the actual widget on the dashboard, so I think the best GA event category to use is "idea_hub_widget". The event names for these are standard: feature_tooltip_view, feature_tooltip_dismiss, feature_tooltip_complete, with the number of the current tooltip as a label.

**Comment by eugene-manuilov** [comment]
> these appear all within the actual widget on the dashboard, so I think the best GA event category to use is "idea_hub_widget". The event names for these are standard: feature_tooltip_view, feature_tooltip_dismiss, feature_tooltip_complete, with the number of the current tooltip as a label.

Thanks, @marrrmarrr, but this is a different feature tour. This one we show on the posts page, not on the dashboard. What do you think if we use the `idea_hub_posts_page` category?

**Comment by felixarntz** [comment]
@eugene-manuilov Let's use `idea_hub_posts` for here.

**Comment by eugene-manuilov** [comment]
@aaemnnosttv @felixarntz this is ready for another round of code review.

**Comment by wpdarren** [comment]
### QA Update: ⚠️ 

@ivankruchkoff since Eugene is out this week, I'd appreciate your thoughts on why the tour is not appearing for me.  

- I am connected to elasticpress.io on a new jurassic.ninja site. 
- Idea hub is active and setup, and a draft post is created.  

When I go to the posts screen no tour tip appears.  Any ideas?

**Comment by ivankruchkoff** [comment]
@wpdarren, here's what we need to test:

1. idea hub active and setup.
2. draft post created
3. tours not yet dismissed, with the testing plugin ( `Site Kit by Google – Test Idea Hub connection` ) you can visit `/wp-admin/?idea-hub-reset`
4. open the post edit screen `/wp-admin/edit.php` and you should see the feature tour.

When I went through these steps, here's what my feature tour looked like:
<img width="1243" alt="Screen Shot 2021-08-24 at 11 10 11 AM" src="https://user-images.githubusercontent.com/373931/130572951-9996e916-d28a-4e7b-8379-9867fed8291d.png">


if the feature tour isn't showing up, a couple of things to check via console to debug (on the post edit screen from step 4 above):

- is the container for the js app available? we can test this via `document.getElementById( 'js-googlesitekit-post-list' );`
- is an idea hub draft post visible? we can test this via `document.querySelector( '.googlesitekit-idea-hub__post' )`

**Comment by wpdarren** [comment]
@ivankruchkoff I'm still not seeing the feature tour on WordPress posts page.  I have gone through the instructions a few times.  When I run the two commands in console this is what I see.  Is this what you expect to see? 

Would it be okay to jump on a quick call. I'll set up a new jurassic.ninja site and screen share with you. 

```
document.getElementById( 'js-googlesitekit-post-list' );
<div id=​"js-googlesitekit-post-list" class=​"googlesitekit-plugin">​</div>​
```

![image](https://user-images.githubusercontent.com/73545194/130615620-f5a062a5-a1ef-4e84-81e0-10406ca4d720.png)

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@ivankruchkoff thank you for your support on this. When I tested it on your live site, it worked straightaway.  I do have an observation though.  The feature tour styling seems to be different than Idea hub.  

This is the font style for the title and description on Idea Hub widget. 

![tours](https://user-images.githubusercontent.com/73545194/130634033-56afe346-6559-4358-a804-ccf28aa1c559.png)

This is how it looks on the Posts page. You can see that the title is not bold.  Also the 'Got it' font should be bold, I suspect.  

![image](https://user-images.githubusercontent.com/73545194/130634196-5394914b-f0e3-415f-a250-60421fab7c65.png)

**Comment by aaemnnosttv** [comment]
@wpdarren – These look _nearly_ the same to me but I see what you're talking about. 

![image](https://user-images.githubusercontent.com/1621608/130659425-d11a3360-b990-4640-9d5c-c9d2c4af14ca.png)
![image](https://user-images.githubusercontent.com/1621608/130659521-7577fc07-4595-4017-9efa-56a3a539a6cd.png)

The different that I do see is that a different font variant is being rendered for the same weight though. On the SK dashboard, the tooltip content is using Roboto Light while on the posts page, it is using Roboto Thin 🤯 

![image](https://user-images.githubusercontent.com/1621608/130660150-b9d93466-2108-476d-9755-ef9572d66f3d.png)
![image](https://user-images.githubusercontent.com/1621608/130660158-e42b8ff9-43b8-43d6-ad69-4e3a7c01c258.png)

**TL;DR** our font's aren't being loaded at all on the post list page due to the on-the-fly loading we do here which tries to use a hook which has already passed. cc: @felixarntz

**Comment by aaemnnosttv** [comment]
@ivankruchkoff @wpdarren after discussing with @felixarntz we decided that we'll address this in a follow-up issue that I'll open tomorrow. The font issue here shouldn't block this from completing QA 👍 

Back to you @wpdarren for a final pass.

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

With no posts in draft the feature tour does not appear on the post page. 
When there are draft posts created the feature tour appears as is displayed as per the AC.

![image](https://user-images.githubusercontent.com/73545194/130783969-7520fcd8-d467-40b4-878c-b5e02401f4e1.png)
